feat(KRY-634): Add sidekick local run to support local bolt endpoint

avasani opened the pull request:
This commit introduces support for specifying a local IP address in the SideKick when running in local mode. It allows the BoltRouter to construct the BoltEndpoints using the local IP address and points both main and failover endpoints to the same port (9000).

This would be useful to these locally since bolt and sidekick would be running on samehost.

Unit Test:
- [x] Run Sidekick in local mode
```
$ docker run -p 7075:7075 --env GRANICA_CUSTOM_DOMAIN=localhost --env AWS_REGION="us-east-1" -v ~/.aws/:/root/.aws/ sidekick:latest serve -l --bolt-endpoint-override 192.168.1.183:9000 -f
```

- [x]  Validate S3 access
```
$ aws s3api list-buckets    --endpoint-url http://localhost:7075/
{
    "Buckets": [
        {
            "Name": "n-data-in1",
            "CreationDate": "2023-07-18T14:16:05.360000+00:00"
        },
        {
            "Name": "in1",
            "CreationDate": "2023-07-18T14:16:05.364000+00:00"
        },
        {
            "Name": "n-virtual-recyclebin",
            "CreationDate": "2023-07-17T21:25:02.407000+00:00"
        }
    ],
    "Owner": {
        "DisplayName": "s3server",
        "ID": "fe7272ea58be830e56fe1663b10fafef"
    }
}
```

kmushegi commented on cmd/serve.go:
```diff
@@ -23,6 +23,7 @@ func init() {
 func initServerFlags(cmd *cobra.Command) {
 	cmd.Flags().IntP("port", "p", DEFAULT_PORT, "The port for sidekick to listen on.")
 	cmd.Flags().BoolP("local", "l", false, "Run sidekick in local (non cloud) mode. This is mostly use for testing locally.")
+	cmd.Flags().StringP("localipaddress", "i", "", "Use local ipaddr to get the local bolt address.")
```
Consider giving this a more descriptive name? it is a address of the local data cruncher instance right? and is used to override the Bolt endpoints so perhaps something like `bolt-endpoint-override` or `bolt-endpoint`? Ditto this in the help message - we're forcing the ip address we pass here to be the Bolt address, I don't think that's super clear at the moment.

Also consider adding dashes vs concatenating 3 words for ux & readability - https://www.gnu.org/software/libc/manual/html_node/Argument-Syntax.html

Given this is public & cx-facing I think we should be mindful of these details.

avasani replied:
I agree, let's go with bolt-endpoint-override.

kmushegi commented on boltrouter/bolt_endpoints.go:
```diff
@@ -73,7 +76,14 @@ func (br *BoltRouter) RefreshEndpoints(ctx context.Context) error {
 	// If running locally, we can't connect to quicksilver.
 	if br.config.Local {
-		return BoltEndpointsMap{}, nil
+		// For local run, return the BoltEndPoints built from localIP address and
+		// main & failover point to same port (9000).
+		endpoints := make(BoltEndpointsMap)
+		endpoints["main_write_endpoints"] = []string{fmt.Sprintf("%s:9000", br.config.LocalIpAddress)}
+		endpoints["failover_write_endpoints"] = []string{fmt.Sprintf("%s:9000", br.config.LocalIpAddress)}
+		endpoints["main_read_endpoints"] = []string{fmt.Sprintf("%s:9000", br.config.LocalIpAddress)}
+		endpoints["failover_read_endpoints"] = []string{fmt.Sprintf("%s:9000", br.config.LocalIpAddress)}
+		return endpoints, nil
```
What happens if `br.config.Local` is true but the user did not pass in the `localipaddress`? We will return `:9000` - would a sensible default like `localhost` make sense for `localipaddress?`

avasani replied:
I think it's better to check if user pass' localipaddress, if not then just 'return BoltEndpointsMap{}'. bolt and SK run on separate container, hence we can not use localhost in this case.

kmushegi replied:
sounds good to me!

kmushegi commented:
Also peep the readme for conventional commits formatting: https://github.com/project-n-oss/sidekick#commits

kmushegi commented on cmd/serve.go:
```diff
@@ -23,6 +23,7 @@ func init() {
 func initServerFlags(cmd *cobra.Command) {
 	cmd.Flags().IntP("port", "p", DEFAULT_PORT, "The port for sidekick to listen on.")
 	cmd.Flags().BoolP("local", "l", false, "Run sidekick in local (non cloud) mode. This is mostly use for testing locally.")
+	cmd.Flags().StringP("bolt-endpoint-override", "i", "", "Specify the local bolt endpoint to override in local mode.")
```
nit: why `i`?

avasani replied:
aah!! missed that. Let me remove that.

dskart commented:
Looks good but make sure that your commits follow the [https://www.conventionalcommits.org/en/v1.0.0/](https://www.conventionalcommits.org/en/v1.0.0/) format!

avasani commented:
> Looks good but make sure that your commits follow the https://www.conventionalcommits.org/en/v1.0.0/ format!

Fixed. I also rebased the code with @kmushegi 's changes.

kmushegi reviewed (approved):
lgtm - I think you'll need to rebase once more!
